server/blob/blob-s3/src/main/java/org/apache/james/blob/objectstorage/aws/sse/S3SSECustomerKeyFactory.java

```diff
@@ -51,4 +52,20 @@ public Mono<SSECustomerKey> generate(BucketName bucketName, BlobId blobId) {
             return Mono.just(sseCustomerKey);
         }
     }
+
+    class PerObjectCustomerKeyFactory implements S3SSECustomerKeyFactory {
```

Comment:
What is the idea behind this class name? I expected a mention of the derived key given that the class takes `S3SSECConfiguration.KeyDerivationEnabled` as an argument.

Reply:
- `SingleCustomerKeyFactory`: uses a "fixed/constant" master key and salt from the configuration file for a single customer key at runtime (apply for all object).
- `PerObjectCustomerKeyFactory`: generates a unique customer key for each object, with the master key as bucketName and salt as blobId, as seen in the method `generate(BucketName bucketName, BlobId blobId)`.

Reply:
> I expected a mention of the derived key 

Derived key = customer key + customer key md5 + algorithm. 
This logic is implemented in the `generate` method.

Reply:
If this class name is confusing, could you suggest a better one?

Reply:
How about?
- ConfiguredCustomerKeyFactory
- DerivedCustomerKeyFactory or S3ObjectDerivedCustomerKeyFactory

Reply:
DerivedCustomerKeyFactory would make sense IMO

Reply:
Will renaming to `DerivedCustomerKeyFactory` in next fixup commit